src/main/scala/org/broadinstitute/dsde/rawls/expressions/SlickExpressionParsing.scala

```diff
@@ -280,38 +307,111 @@ trait SlickExpressionParser extends JavaTokenParsers {
       entity <- entityQuery if attribute.valueEntityRef === entity.id
     } yield entity
 
-    val entitiesAction = entityQuery.unmarshalEntities(entityQuery.joinOnAttributesAndRefs(query))
-    entitiesAction.map { case (entities) => sortEntities(entities) }
+    //Return the value of the expression once for each entity we wanted to evaluate this expression against!
+    query.sortBy(_.name.asc).result map { resultEntities =>
+      (for {
+        entity <- context.rootEntities
+      } yield (entity.name, resultEntities)).toMap
+    }
   }
 
-  private def stringFunc(str: String)(context: SlickExpressionContext, shouldBeNone: Option[PipeType]): ReadAction[Iterable[Attribute]] = {
+  private def stringFunc(str: String)(context: SlickExpressionContext, shouldBeNone: Option[PipeType]): ReadAction[Map[String,Iterable[Attribute]]] = {
     assert(shouldBeNone.isEmpty)
-    DBIO.successful(Seq(AttributeString(str.drop(1).dropRight(1))))
+    DBIO.successful( context.rootEntities.map( _.name -> Seq(AttributeString(str.drop(1).dropRight(1)))).toMap )
   }
 
-  private def floatFunc(dec: String)(context: SlickExpressionContext, shouldBeNone: Option[PipeType]): ReadAction[Iterable[Attribute]] = {
+  private def floatFunc(dec: String)(context: SlickExpressionContext, shouldBeNone: Option[PipeType]): ReadAction[Map[String,Iterable[Attribute]]] = {
     assert(shouldBeNone.isEmpty)
-    DBIO.successful(Seq(AttributeNumber(BigDecimal(dec))))
+    DBIO.successful( context.rootEntities.map( _.name -> Seq(AttributeNumber(BigDecimal(dec)))).toMap )
   }
 }
 
-class SlickExpressionEvaluator(parser: SlickExpressionParser)(implicit executionContext: ExecutionContext)  {
-  def evalFinalAttribute(workspaceContext: SlickWorkspaceContext, rootType:String, rootName:String, expression:String): Try[ReadAction[Iterable[AttributeValue]]] = {
-    parser.parseAttributeExpr(expression).map( runPipe(SlickExpressionContext(workspaceContext, rootType, rootName), _).map(_.collect {
-      case AttributeNull => Seq.empty
-      case AttributeEmptyList => Seq.empty
-      case av: AttributeValue => Seq(av)
-      case avl: AttributeValueList => avl.list
-      case badType => throw new RawlsException(s"unsupported type resulting from attribute expression: $badType: ${badType.getClass}")
-    }.flatten))
+object SlickExpressionEvaluator {
+  def withNewExpressionEvaluator[R](parser: DataAccess, rootEntities: Seq[EntityRecord])
+                                   (op: SlickExpressionEvaluator => ReadWriteAction[R])
+                                   (implicit executionContext: ExecutionContext): ReadWriteAction[R] = {
+    val evaluator = new SlickExpressionEvaluator(parser, rootEntities)
+
+    evaluator.populateExprEvalTempTable()
+      .andThen({ op(evaluator) })
+      .cleanUp({ _ => evaluator.dropExprEvalTempTable() })
```

Comment:
I wonder if it is a good idea to run each expression in its own transaction. Benefits: don't need to clean up, possibly performance boost because they can run concurrently. Pitfalls: 1 user can use possibly many of our db connections starving other operations.